Sources/Core/Internal/PluginManager.swift

```diff
@@ -12,16 +12,19 @@ internal final class PluginManager: Activatable {
 
     // MARK: - API
 
-    func add<P: Plugin>(_ pluginProvider: () -> P, for object: P.Object) {
+    func add<P: Plugin>(_ pluginProvider: () -> P, for object: P.Object) -> P {
         let identifier = ObjectIdentifier(P.self)
 
-        guard plugins[identifier] == nil else {
-            return
+        if let existingPlugin = plugins[identifier] {
+            return existingPlugin.wrapped as! P
```

Comment:
It's a little misleading that the function is called `add` but really it is adding or retrieving (and ignoring the passed in plugin). That is not what I'd expect calling `add` on a `PluginManager` with a new instance of the plugin to do, I'd expect the new instance to replace the previous. I think there at least needs to be documentation added to point out this possibly unexpected behavior, and maybe even consider a rename of the function? Though I really don't have a good suggestion for a different name...

Reply:
Yeah I totally agree 👍

I think I have an idea how to solve this 😄 How about we make the "only add if there's not already a plugin of this class" behavior controlled by a parameter? That way we could make the signature a lot more clear, like this:

```swift
func add<P: Plugin>(_ plugin: P, reuseExistingOfSameType: Bool = true) -> P
```

That way it becomes a lot more self-documenting why the method is returning something, and why it might be a different instance than the one passed in in the auto closure. What do you think @aranasaurus?

Reply:
Yeah, that sounds great! 👍